### ISSUE TITLE
Modify `FlowOps.expand` to better match its doc

### ISSUE THREAD

**Issue by mikolak-net** [akka/akka#23804]
I would propose that the current implementation of [`expand`](https://doc.akka.io/api/akka/2.5.6/akka/stream/scaladsl/Flow.html#expand[U](extrapolate:Out=%3EIterator[U]):FlowOps.this.Repr[U]) has two gotchas:

1. It is unclear how the `extrapolate` argument should be used. From the first sentence of the doc one might imply that the iterator is invoked *only* when the downstream signals demand, and not for the original element itself. I.e. I'd argue that this:

```scala
  implicit val system = ActorSystem()
  implicit val m = ActorMaterializer()

  val resultList = Source(List("1", "2", "3", "4")).expand(_ => Iterator("a")).runWith(Sink.seq)

  println(Await.result(resultList, Duration.Inf))
```

Should print out `Vector(1, 2, 3, 4)`, whereas the actual printout is `Vector(a, a, a, a)`.

This behavior does not appear to provide any functionality that couldn't be replicated with a combination of `map` and `splitX` steps (and arguably in a more efficient fashion).

_(and, to further flog that dead horse, please see the [dictionary definition](https://en.wiktionary.org/wiki/extrapolate) of "extrapolate" - in fact, in an informal poll among several fellow developers, all were tripped up by the aforementioned issue)_

2. There's no possibility to handle the "degenerate" case of the downstream signaling demand for the pipeline while no elements are present. Serendipitously, there's actually a provision for that in the current (2.5.6) Scaladoc:
```scala
@param seed Provides the first state for extrapolation using the first unconsumed element
```

but it's not reflected in the code.

-------

To summarize, I would propose for `FlowOps.expand` to have either one of the following definitions:

```scala
def expand(extrapolate: Out ⇒ Iterator[Out], seed: Option[Out] = None): Repr[Out]
```

or

```scala
def expand(extrapolate: Out ⇒ Iterator[Out], seed: Out): Repr[Out]
```

where `extrapolate` would *only* be used whenever nothing can be `pull` ed from upstream on demand.

(note the omission of the `U` parameter, to remove the possibility of abusing `expand` as a `map` with additional bells and whistles)

**Comment by johanandren** [comment]
Thanks for the writeup. A mixup of downstream and upstream here, made it a bit hard to follow at first - downstream is where demand/backpressure comes from upstream is whee you pull.

I don't think the sample to prove that the behavior is strange makes sense at all, something that would better show how expand works is this:
```
  val resultList = Source(List("a", "b", "c", "d"))
    .expand(c => Iterator.from(1).map(n => n + c))
    .runForeach(println)
```

Which will print something like: 
```
1a
2a
1b
1c
1d
```

Where `1a` and `2a` are examples of the extrapolation.

However, I think an optional seed make sense, we'd have to keep the old signature though, for binary compatibility.

Up for making a PR with this?

**Comment by mikolak-net** [comment]
@johanandren : re upstream/downstream - gah, sorry about that! Fixed the OP.

Re semantics - I'm not sure I've articulated myself correctly here. The point of confusion/unclarity in my opinion comes from that the fact that the `extrapolate` argument always modifies the original element. So, to use your example's output, I would expect the return to be:
```
a
1a
b
c
d
```

And this would be sufficient for downstream stage to recognize the synthetic element.

Like I said, I don't really see a use case where a combination of other stages wouldn't be a better fit. In fact, the "repeat" example from the docs would function identically under the proposed change. Could you provide an example where the current semantics of `expand` is necessary, given the context?

Re binary compatibility - fair point.

Re PR - of course, once we reach a consensus :).

**Comment by johanandren** [comment]
Hmm. Ok. That is fairly explicitly noted in the docs though: 

> This element will never "drop" upstream elements as all elements go through at least one extrapolation step.

Achieving your expected output is easy, just make sure you use the incoming element as first extrapolated:
```
  val resultList = Source(List("a", "b", "c", "d"))
    .expand(c => Iterator(c) ++ Iterator.from(1).map(n => n + c))
    .runForeach(println)
```

I think changing the semantic would be dangerous as the current ones has been out a couple of years and there is likely code depending on the behavior out there. If anything should be done about that it would be to update documentation to be even more explicit about the `extrapolate` function always being applied.

Where I see we can change would be by providing an extra parameter for an optional first seed to extrapolate from, that is only ever used for extrapolation if upstream has not emitted a first element yet.

What the `expand` provides over `map`and `splitX` combinations that you mention is that it avoids submaterializations, so it is more efficient/performant.

**Comment by mikolak-net** [comment]
Like I said, the use of "extrapolate" in this place is at odds with what "extrapolation" means (one "extrapolates Y from X", an not "X from X"). For me this reads somewhat nonsensically - and seeing as how "extrapolate" is a Latin loanword, I doubt it's a language barrier thing.

To clarify, my intention is not to nitpick about lexicographic errors - it is to highlight that the API is misleading in its current form (and that's why I made the aforementioned impromptu poll among my peers - I had literally a sample of 4, there was nevertheless a 100% error rate, so that should give something to think about).

Also, note that this piece of the doc is factually incorrect - providing `expand` with something returning `Iterator.empty` *will* make the stage drop the relevant elements - yet another gotcha'.

Regarding `map`/`splitX` - that was for the general case. In all the examples discussed here, you could replicate the functionality with non-subflow stages, e.g. adding `map` before/after `expand`.

I agree however that changing the semantic at this point of this specific `expand` method would be dangerous.

In the interest avoiding a drawn-out discussion, how about if I start with the PR and we'll see whether whatever I cooked will be (eventually) mergeable?

**Comment by johanandren** [comment]
As I mentioned before, the stage has had the same naming since the initial experimental release of streams in July 2015 and this is the first time there is any discussions around it as far as I can remember mailing-list/gitter channel and issue tracker. The only related earlier related thing I can find is the issue from when the initial (!!) seed parameter was removed (#19444)

Scala has support for some form of parameter name-change-deprecation that I didn't think of before, I can't remember if that is an issue with binary compatibility but I don't think so, and it should solve the source compatibility side of thing. That together with adding some caveats to the docs, for example about returning an iterator that is empty seems like the most bang for the bucks to me.

Another option would perhaps introduce a new stage/factory with a name and and parameter that is more fitting/self explanatory/has less surprising semantics (that can not be `expand` then, because that would be really confusing) and maybe but not necessarily deprecate the current `expand`. 

WDYT?

**Comment by mikolak-net** [comment]
I very much understand your reservations - I realize the claims I'm making are quite bold. Like I said, I wouldn't be so forward if my informal "tests" haven't pointed towards the ambiguity/unclarity being there.

It looks we're roughly on the same page now. I totally agree re confusion, so I'm going to start with introducing a completely new (name-wise) method in the PR and we'll work out a consensus from there. Sounds good?

**Comment by johanandren** [comment]
I'll add a 👍 here as maybe just doing that on the comment doesn't trigger an event :)

**Comment by mikolak-net** [comment]
@johanandren : heh, yeah, I was beginning to wonder about the "lack" of response and was getting to bumping the issue :). Great, onto the PR then!

**Comment by mikolak-net** [comment]
Update - still up for it, might take me 'till the end of December to cook up the PR 'though due to other obligations.

**Comment by mikolak-net** [comment]
Another heads-up, I finally have some free time to set aside for this, I'm hoping for a WIP PR by the end of the week.

**Comment by ktoso** [comment]
Looking forward to it, thanks :)
